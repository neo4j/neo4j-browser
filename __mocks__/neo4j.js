/*
 * Copyright (c) 2002-2018 "Neo4j, Inc"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

function integerFn (val) {
  this.val = val
}
integerFn.prototype.toString = function () {
  return this.val.toString()
}

var out = {
  v1: {
    Int: integerFn,
    isInt: function (val) {
      return val instanceof integerFn
    },
    types: {
      Node: function Node (id, labels, properties) {
        this.identity = id
        this.labels = labels
        this.properties = properties
      },
      Relationship: function Relationship (id, start, end, type, properties) {
        this.identity = id
        this.start = start
        this.end = end
        this.type = type
        this.properties = properties
      },
      Path: function Path (start, end, segments) {
        this.start = start
        this.end = end
        this.segments = segments
        this.length = segments.length
      },
      PathSegment: function PathSegment (start, relationship, end) {
        this.start = start
        this.relationship = relationship
        this.end = end
      }
    },
    Integer: function Integer ({ low, high }) {
      this.low = low
      this.high = high
    }
  }
}

out.v1.types.Node.prototype.toString = function () {
  return 'node'
}
out.v1.types.Relationship.prototype.toString = function () {
  return 'rel'
}
out.v1.types.Path.prototype.toString = function () {
  return 'path'
}
out.v1.types.PathSegment.prototype.toString = function () {
  return 'pathsegment'
}
out.v1.Integer.prototype.toInt = function () {
  return this.low
}
out.v1.int = val => new out.v1.Integer(val)

module.exports = out
