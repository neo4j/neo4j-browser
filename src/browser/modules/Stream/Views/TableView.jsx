/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Component } from 'preact'
import { v4 } from 'uuid'
import { PaddedDiv, StyledBodyMessage } from '../styled'
import {StyledTable, StyledBodyTr, StyledTh, StyledTd} from 'browser-components/DataTables'
import { stripQuotes } from 'shared/services/utils'

class TableView extends Component {
  constructor (props) {
    super(props)
    const dataCopy = props.data ? props.data.slice() : []
    const headerData = dataCopy.length > 0 ? dataCopy.shift() : []
    this.state = {
      columns: headerData,
      data: dataCopy
    }
  }
  render () {
    if (!this.props.data) return (<PaddedDiv style={this.props.style}><StyledBodyMessage>{this.props.message}</StyledBodyMessage></PaddedDiv>)
    const tableHeader = this.state.columns.map((column, i) => (
      <StyledTh className='table-header' key={i}>{stripQuotes(column)}</StyledTh>)
    )
    const buildData = (entries) => {
      return entries.map((entry) => {
        if (entry !== null) {
          return <StyledTd className='table-properties' key={v4()}>{entry}</StyledTd>
        }
        return <StyledTd className='table-properties' key={v4()}>(empty)</StyledTd>
      })
    }
    const buildRow = (item) => {
      return (
        <StyledBodyTr className='table-row' key={v4()}>
          {buildData(item)}
        </StyledBodyTr>
      )
    }
    const tableBody = (
      <tbody>
        {
          this.state.data.map((item) => (
            buildRow(item)
          ))
        }
      </tbody>
    )
    return (
      <PaddedDiv style={this.props.style}>
        <StyledTable>
          <thead>
            <tr>
              {tableHeader}
            </tr>
          </thead>
          {tableBody}
        </StyledTable>
      </PaddedDiv>
    )
  }
}

export default TableView
