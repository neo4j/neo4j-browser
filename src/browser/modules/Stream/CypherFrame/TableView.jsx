/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { v4 } from 'uuid'
import neo4j from 'neo4j-driver'
import { sanitize } from 'dompurify'

import {
  StyledStatsBar,
  PaddedTableViewDiv,
  StyledBodyMessage
} from '../styled'
import Ellipsis from 'browser-components/Ellipsis'
import {
  StyledTable,
  StyledBodyTr,
  StyledTh,
  StyledTd,
  StyledJsonPre
} from 'browser-components/DataTables'
import { shallowEquals, stringifyMod } from 'services/utils'
import {
  getBodyAndStatusBarMessages,
  getRecordsToDisplayInTable,
  transformResultRecordsToResultArray
} from './helpers'
import { stringModifier } from 'services/bolt/cypherTypesFormatting'
import ClickableUrls, {
  convertUrlsToHrefTags
} from '../../../components/clickable-urls'

const renderCell = entry => {
  if (Array.isArray(entry)) {
    const children = entry.map((item, index) => (
      <span key={index}>
        {renderCell(item)}
        {index === entry.length - 1 ? null : ', '}
      </span>
    ))
    return <span>[{children}]</span>
  } else if (typeof entry === 'object') {
    return renderObject(entry)
  } else {
    return <ClickableUrls text={stringifyMod(entry, stringModifier, true)} />
  }
}
export const renderObject = entry => {
  if (neo4j.isInt(entry)) return entry.toString()
  if (entry === null) return <em>null</em>
  return (
    <StyledJsonPre
      dangerouslySetInnerHTML={{
        __html: convertUrlsToHrefTags(
          sanitize(stringifyMod(entry, stringModifier, true))
        )
      }}
    />
  )
}
const buildData = entries => {
  return entries.map(entry => {
    if (entry !== null) {
      return (
        <StyledTd className="table-properties" key={v4()}>
          {renderCell(entry)}
        </StyledTd>
      )
    }
    return (
      <StyledTd className="table-properties" key={v4()}>
        null
      </StyledTd>
    )
  })
}
const buildRow = item => {
  return (
    <StyledBodyTr className="table-row" key={v4()}>
      {buildData(item)}
    </StyledBodyTr>
  )
}

export class TableView extends Component {
  state = {
    columns: [],
    data: [],
    bodyMessage: ''
  }

  componentDidMount() {
    this.makeState(this.props)
  }

  componentWillReceiveProps(props) {
    if (
      this.props === undefined ||
      this.props.result === undefined ||
      this.props.updated !== props.updated
    ) {
      this.makeState(props)
    }
  }

  shouldComponentUpdate(props, state) {
    return (
      this.props.updated !== props.updated || !shallowEquals(state, this.state)
    )
  }

  makeState(props) {
    const records = getRecordsToDisplayInTable(props.result, props.maxRows)
    const table = transformResultRecordsToResultArray(records) || []
    const data = table ? table.slice() : []
    const columns = data.length > 0 ? data.shift() : []
    const { bodyMessage } = getBodyAndStatusBarMessages(
      props.result,
      props.maxRows
    )
    this.setState({ data, columns, bodyMessage })
  }

  render() {
    if (!this.state.columns.length) {
      return (
        <PaddedTableViewDiv>
          <StyledBodyMessage>{this.state.bodyMessage}</StyledBodyMessage>
        </PaddedTableViewDiv>
      )
    }
    const tableHeader = this.state.columns.map((column, i) => (
      <StyledTh className="table-header" key={i}>
        {column}
      </StyledTh>
    ))
    const tableBody = (
      <tbody>{this.state.data.map(item => buildRow(item))}</tbody>
    )
    return (
      <PaddedTableViewDiv>
        <StyledTable>
          <thead>
            <tr>{tableHeader}</tr>
          </thead>
          {tableBody}
        </StyledTable>
      </PaddedTableViewDiv>
    )
  }
}

export class TableStatusbar extends Component {
  state = {
    statusBarMessage: ''
  }

  componentDidMount() {
    this.makeState(this.props)
  }

  componentWillReceiveProps(props) {
    this.makeState(props)
  }

  shouldComponentUpdate(props, state) {
    if (!shallowEquals(state, this.state)) return true
    return false
  }

  makeState(props) {
    const { statusBarMessage } = getBodyAndStatusBarMessages(
      props.result,
      props.maxRows
    )
    if (statusBarMessage !== undefined) this.setState({ statusBarMessage })
  }

  render() {
    return (
      <StyledStatsBar>
        <Ellipsis>{this.state.statusBarMessage}</Ellipsis>
      </StyledStatsBar>
    )
  }
}
