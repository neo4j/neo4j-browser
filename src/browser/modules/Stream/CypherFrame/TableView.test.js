/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render } from '@testing-library/react'
import neo4j from 'neo4j-driver'

import { TableView, TableStatusbar, renderObject } from './TableView'

describe('TableViews', () => {
  describe('TableView', () => {
    test('displays bodyMessage if no rows', () => {
      // Given
      const sps = jest.fn()
      const props = {
        setParentState: sps,
        result: {
          records: [],
          summary: {
            resultAvailableAfter: neo4j.int(5),
            resultConsumedAfter: neo4j.int(5)
          }
        }
      }

      // When
      const { container } = render(<TableView {...props} />)

      // Then
      expect(container).toMatchSnapshot()
    })
    test('does not display bodyMessage if rows', () => {
      // Given
      const sps = jest.fn()
      const result = {
        records: [{ keys: ['x'], _fields: ['y'], get: () => 'y' }],
        summary: {
          resultAvailableAfter: neo4j.int(5),
          resultConsumedAfter: neo4j.int(5)
        }
      }

      // When
      const { container } = render(
        <TableView setParentState={sps} result={result} />
      )

      // Then
      expect(container).toMatchSnapshot()
    })
    test('renderObject handles null values', () => {
      // Given
      const datas = [{ x: 1 }, null]

      // When
      const results = datas.map(data => renderObject(data))

      // Then
      results.forEach((res, i) => expect(res).toMatchSnapshot())
    })
  })
  describe('TableStatusbar', () => {
    test('displays no statusBarMessage', () => {
      // Given
      const props = { result: {}, maxRows: 0 }

      // When
      const { container } = render(<TableStatusbar {...props} />)

      // Then
      expect(container).toMatchSnapshot()
    })
    test('displays statusBarMessage', () => {
      // Given
      const props = {
        query: 'MATCH xx0',
        status: 'success',
        result: {
          summary: {
            resultAvailableAfter: neo4j.int(5),
            resultConsumedAfter: neo4j.int(5)
          },
          maxRows: 100,
          records: [{ res: 'xx3' }]
        }
      }

      // When
      const { container } = render(<TableStatusbar {...props} />)

      // Then
      expect(container).toMatchSnapshot()
    })
  })
})
