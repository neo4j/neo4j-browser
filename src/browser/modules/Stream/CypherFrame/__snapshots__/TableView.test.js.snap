// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableViews TableStatusbar displays no statusBarMessage 1`] = `
<div>
  <div
    class="styled__StyledStatsBar-sc-1dtvgs1-32 hNrxzW"
  >
    <div
      class="Ellipsis-sc-1dn9bou-0 gzqRPC"
    />
  </div>
</div>
`;

exports[`TableViews TableStatusbar displays statusBarMessage 1`] = `
<div>
  <div
    class="styled__StyledStatsBar-sc-1dtvgs1-32 hNrxzW"
  >
    <div
      class="Ellipsis-sc-1dn9bou-0 gzqRPC"
    >
      Started streaming 1 records after 5 ms and completed after 10  ms.
    </div>
  </div>
</div>
`;

exports[`TableViews TableView displays bodyMessage if no rows 1`] = `
<div>
  <div
    class="styled__PaddedDiv-sc-1dtvgs1-6 styled__PaddedTableViewDiv-sc-1dtvgs1-7 iNSchp"
  >
    <div
      class="styled__StyledBodyMessage-sc-1dtvgs1-35 hfUgGP"
    >
      (no changes, no records)
    </div>
  </div>
</div>
`;

exports[`TableViews TableView does not display bodyMessage if rows 1`] = `
<div>
  <div
    class="styled__PaddedDiv-sc-1dtvgs1-6 styled__PaddedTableViewDiv-sc-1dtvgs1-7 iNSchp"
  >
    <table
      class="DataTables__StyledTable-bf850j-0 cZayOW"
    >
      <thead>
        <tr>
          <th
            class="table-header DataTables__StyledTh-bf850j-2 gjWFrA"
          >
            x
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="table-row DataTables__StyledBodyTr-bf850j-1 iGjwRr"
        >
          <td
            class="table-properties DataTables__StyledTd-bf850j-3 jDsjVi"
          >
            "y"
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`TableViews TableView renderObject handles null values 1`] = `
<ForwardRef>
  {
  "x": 1.0
}
</ForwardRef>
`;

exports[`TableViews TableView renderObject handles null values 2`] = `
<em>
  null
</em>
`;
